Add support for Ring Sirens and floodlights

rossdargan opened the pull request:
## Description:
Allows a user to turn on the siren for all Ring Stickup cams that support it (Tested with a Stick up cam and a Floodlight cam). 

Allows a user to turn on the flood lights for all cameras that support it (Tested on a Floodlight cam)

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** https://github.com/home-assistant/home-assistant.io/pull/9948

## Example entry for `configuration.yaml` (if applicable):
```yaml
switch:
  - platform: ring
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`. **Note** There is no owner here... but I don't feel like the right person to own this!
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`. **NA**
  - [x] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works. **NA**

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

rossdargan committed:
Add Siren and Light support for Ring devices

rossdargan committed:
Add Siren and Light support for Ring devices

rossdargan committed:
add ring/switch.py to .coveragerc

rossdargan committed:
add ring/switch.py to .coveragerc

homeassistant commented:
Hello @rossdargan,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`ecfa362294dee280b256cb995f2c08f57a999b71`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25397&#x2F;commits&#x2F;ecfa362294dee280b256cb995f2c08f57a999b71) - This commit has something that looks like an email address (ross.dargan@waterstons.com). Maybe try linking that to GitHub?.
- [`6ed8c73cf7aeaa31b88c856248d5a00b0a5afcf1`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;25397&#x2F;commits&#x2F;6ed8c73cf7aeaa31b88c856248d5a00b0a5afcf1) - This commit has something that looks like an email address (ross.dargan@waterstons.com). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

rossdargan committed:
fix spacing issues

rossdargan committed:
fix spacing issues

rossdargan commented on homeassistant/components/ring/switch.py:
```diff
@@ -0,0 +1,115 @@
+
+    def update(self):
+        """Get the siren status from ring"""
+        self._device.update()
```
This worries me a bit because for every device I will be making an update call (so if a device has a camera and a floodlight then there will be 2 calls) every 10 seconds. I'm guessing the sensors will be doing similar. I've taken a look at the code this calls (https://github.com/tchellomello/python-ring-doorbell/blob/master/ring_doorbell/generic.py#L52) and there isn't any sort of caching or anything to stop rapid calls here. I'm just not sure of a better way of doing this, or if it's just not a problem I should worry about?

MartinHjelmare replied:
The pattern we normally use, is to move the data fetching to a single data instance that handles all the device or service interactions.

If there's no cache in the library we could build our own cache. Then give access to the data instance with cache to all entities and either let them pull state from the cache via the builtin home assistant polling mechanism or turn of polling and signal the entities to pull state after updating the cache via custom scheduling. If the former builtin polling mechanism is used, there needs to be a throttle on the data instance update method, so that just the first entity that calls the method will actually update the data and cache, within a time interval.

amelchio committed:
Update pysonos to 0.0.22 (#25399)

zxdavb committed:
[climate] Correct evohome hvac_action (#25407)

* inital commit

* take account of rounding up of curr_temp

bachya committed:
Add area support to fan service schemas (#25409)

bachya committed:
Add area support to cover service schemas (#25408)

* Add area support to cover service schemas

* Linting

bachya committed:
Add area support to automation service schemas (#25403)

* Add area support to automation service schemas

* Fixed import

* Mispelling

bachya committed:
Add area support to counter service schemas (#25401)

* Add area support to counter service schemas

* Updates

bachya committed:
Add area support to alarm_control_panel service schemas (#25402)

* Add area support to alarm_control_panel service schemas

* Corrected import

zxdavb committed:
[climate] Bugfix/Tweak honeywell migration (#25369)

* refactor supported_features

add dr_data

delint

simplify code

refactor target_temps, and delinting

delint

fix typo

add min/max temps

delint

refactor target temps

refactor target temps 2

correct typo

tweak

fix potential bug

fix potential bug 2

bugfix entity_id

fix typo

* remove explicit entity_id

* refactor hvac_action

* refactor hvac_action

* bugfix - HVAC_MODE_HEAT_COOL incorrectly added to hvac_modes

cgtobi committed:
Clean up Netatmo sensor code (#25390)

* Clean up code

* Add parameter

* Make it a list

* Move loop and add debug message

* Further clean up

* Move manual config

cgtobi committed:
Bump pyatmo version to 2.1.2 (#25296)

bachya committed:
Add area support to script service schemas (#25439)

bachya committed:
Add area support to scene service schemas (#25438)

bachya committed:
Add area support to input select service schemas (#25432)

bachya committed:
Add area support to input number service schemas (#25431)

bachya committed:
Add area support to input datetime service schemas (#25430)

* Add area support to input datetime service schemas

* Fixed tests

bachya committed:
Add area support to Wink service schemas (#25445)

bachya committed:
Add area support to input boolean service schemas (#25429)

bachya committed:
Add area support to image processing service schemas (#25428)

bachya committed:
Add area support to input text service schemas (#25434)

bachya committed:
Add area support to climate service schemas (#25441)

* Add area support to climate service schemas

* Incorrect Voluptuous usage

* Fix typo

JXGA committed:
Upgrade HPILO requirement to v4.3 (#25444)

* Update Python-HPILO to 4.3

Update of Python-HPILO requirement to 4.3 to resolve outstanding SSL connections for older HP servers (ILO 3)

* Update requirements_all.txt

Update HPILO to 4.3

andre-richter committed:
Increase vallox robustness on startup (#25382)

* Vallox: Increase robustness on startup

Experiments showed that timing of websocket requests to the Vallox firmware is
critical when fetching new metrics. Tests on different Raspberry Pis and x86
machines showed that those machines with little processing power tend to fail
the timing requirments during the busy startup phase of Home Assistant,
resulting in the Vallox integration failing to set itself up.

This patch catches Websocket's InvalidMessage, which is a symptom of failing the
timing requirements. Experiments again showed that on the Raspberry's, this
exception is catched once at startup, but the integration is running fine
afterwards.

* Update __init__.py

* Bump to new 2.1.0 version of api.

* Bump to api 2.2.0

webdjoe committed:
Add Support for VeSync Devices - Outlets and Switches (#24953)

* Change dependency to pyvesync-v2 for vesync switch

* async vesync component

* FInish data_entry_flow

* Update config flow

* strings.json

* Minor fix

* Syntax fix

* Minor Fixs

* UI Fix

* Minor Correct

* Debug lines

* fix device dictionaries

* Light switch fix

* Cleanup

* pylint fixes

* Hassfest and setup scripts

* Flake8 fixes

* Add vesync light platform

* Fix typo

* Update Devices Service

* Fix update devices service

* Add initial test

* Add Config Flow Tests

* Remove Extra Platforms

* Fix requirements

* Update pypi package

* Add login to config_flow

Avoid setting up component if login credentials are invalid

* Fix variable import

* Update config_flow.py

* Update config_flow.py

* Put VS object into hass.data instead of config entry

* Update __init__.py

* Handle Login Error

* Fix invalid login error

* Fix typo

* Remove line

* PEP fixes

* Fix change requests

* Fix typo

* Update __init__.py

* Update switch.py

* Flake8 fix

* Update test requirements

* Fix permission

* Address change requests

* Address change requests

* Fix device discovery indent, add MockConfigEntry

* Fix vesynclightswitch classs

* Remove active time attribute

* Remove time_zone, grammar check

bachya committed:
Add area support to utility meter service schemas (#25442)

* Add area support to utility meter service schemas

* Reverted mistaken deletion

pattyland committed:
Add Elgato Avea integration (#24281)

* Adds Elgato Avea integration

* Revert "Adds Elgato Avea integration"

This reverts commit 8607a685eb47e99ea0e8a69fbdb84e511aaf1be0.

* Adds Elgato Avea integration

* Removed debug

* Improved readability

Co-Authored-By: Otto Winter <otto@otto-winter.com>

* Adds Elgato Avea integration

* Fixes for flake8

* More Fixes for flake8

* Hopefully last fixes for flake8

* Unnecessary rounding and typo removed

* Duplicate calls removed

* raise PlatformNotReady if communication with bulb failes

* Fixes: flake8, missing import of exception and better handling of ble problems

* Update requirements_all.txt

Add requirements_all.txt file

* Revert "Update requirements_all.txt"

This reverts commit 2856025ed3d88bb081e3a8d0fb065f1e7ba1ef45.

* Update requirements_all.txt

* conform with snake_case naming style

https://circleci.com/gh/home-assistant/home-assistant/31823

* Fixed variable rename

* Unnecessary calculation removed

Co-Authored-By: Otto Winter <otto@otto-winter.com>

* Better Exception Handling

* Changed position of import, renamed add_entities to add_devices, remove unnecessary comment

* Unnecessary comments removed.

bachya committed:
Add area support to lock service schemas (#25435)

* Add area support to lock service schemas

* extend

* Fixed tests

* Fixed tests

bachya committed:
Add area support to remote service schemas (#25437)

* Add area support to remote service schemas

* Base

ooii committed:
Suez water (#23844)

* Add suez water sensor

* flake8 test

* pylint test

* edition to fix flake8 and pylint issues

* edition to be okay with the musts

* Added a blank line to __init.py__ for flake8

* added blank line for flake8

* changer scan interval from 10 to 720 minutes

* use of pysuez

* bug fix and isort

* use of pysuez

* fixed flake8 and pylint errors

* update requirements_all.txt

* added a method to test login/password befire adding device

* flake8 edition

* update requirements_all.txt

* add of .coveragerc file with untested files

* update of .coveragerc

* Update homeassistant/components/suez_water/__init__.py

Co-Authored-By: Fabian Affolter <mail@fabian-affolter.ch>

* Update homeassistant/components/suez_water/sensor.py

Co-Authored-By: Fabian Affolter <mail@fabian-affolter.ch>

* Update homeassistant/components/suez_water/sensor.py

Co-Authored-By: Fabian Affolter <mail@fabian-affolter.ch>

* Update homeassistant/components/suez_water/sensor.py

Co-Authored-By: Fabian Affolter <mail@fabian-affolter.ch>

* Update homeassistant/components/suez_water/sensor.py

Co-Authored-By: Fabian Affolter <mail@fabian-affolter.ch>

* Update homeassistant/components/suez_water/sensor.py

Co-Authored-By: Fabian Affolter <mail@fabian-affolter.ch>

* bug fix in check credentials

* flake8 and pylint fixes

* fix codeowner

* update requirements_all.txt

* Sorted suez_water line

* edition to answer comments from @MartinHjelmare on #23844

* Attribute keys formatting to lowercase snakecase, name and icon constants returned directly, and remove of  attribute. Update of .

* pylint edition

* correction wrong keys in client attributes

* remove of unnedeed return and move add_entities

kimfrellsen committed:
Add new device tracker supporting Fortinet FortiGate (#23078)

* Create device_tracker.py

initial version

* Update device_tracker.py

set verify SSL to false as default. Normally users do not have a verified certificate at home

* Update device_tracker.py

pep8 compliant

* Update device_tracker.py

upgraded fortiosapi requirements

* Create __init__.py

* tox compliant

* Update device_tracker.py

* Create manifest.json

* Update .coveragerc

added fortios

* Update device_tracker.py

circle ci, blank line required

* Update manifest.json

removed code owners

* Update manifest.json

removed dependencies

* Update manifest.json

removed codeowners

* Update requirements_all.txt

added fortios

* Update requirements_all.txt

* Update device_tracker.py

pylint corrections

* Update device_tracker.py

pylint exceptions

* Update device_tracker.py

disable pylint broad exceptions

* Update device_tracker.py

pylint

* Update device_tracker.py

removed pointless string statements

* Update device_tracker.py

removed blank line

* Update device_tracker.py

* Update device_tracker.py

* Update device_tracker.py

* Update device_tracker.py

* Update device_tracker.py

* Update device_tracker.py

* Update device_tracker.py

* Update device_tracker.py

* Update device_tracker.py

* Update device_tracker.py

* Update device_tracker.py

* Update device_tracker.py

* Update device_tracker.py

* Update manifest.json

added codeowners

* Update CODEOWNERS

added kimfrellsen as codeowner

* fortiosapi 0.10.8

Updated to use latest version of fortiosapi 0.10.8

* Update requirements_all.txt

updated fortiosapi to 0.10.8

* Update device_tracker.py

fixed some requests.

* Update device_tracker.py

better exception handling.

* Update device_tracker.py

exception handling

* Update CODEOWNERS

* Update device_tracker.py

corrected exception handling

* Update device_tracker.py

exception handling.

* Update device_tracker.py

lint corrections

* Update device_tracker.py

removed broad exception.

* Update device_tracker.py

fix lint errors

* Update device_tracker.py

minor changes, mostly cosmetic

fredrike committed:
Update Daikin preset modes (#25395)

* fix preset documentation

* Use pydaikin set holiday method

Adminiuga committed:
Bump up ZHA dependencies. (#25450)

bachya committed:
Add area support to media player service schemas (#25436)

* Add area support to media player service schemas

* Re-establish MEDIA_PLAYER_SCHEMA

* Comment

* Localize platforms that used MEDIA_PLAYER_SCHEMA

scop committed:
Huawei LTE misc improvements (#25377)

* Rename traffic_statistics to monitoring_traffic_statistics

For better consistency with huawei-lte-api.

* Add default device name for sensors

In case the actual device name cannot be accessed for some reason.

* Support device class in sensor metadata

* Mark known signal strength sensors as such

pvizeli committed:
Update azure-pipelines-release.yml for Azure Pipelines

bachya committed:
Add area support to timer service schemas (#25440)

* Add area support to timer service schemas

* Base

* Remove unnecessary TIMER_SERVICE_SCHEMA

bachya committed:
Remove unnecessary REMOTE_SERVICE_SCHEMA (#25453)

bachya committed:
Add area support to vacuum service schemas (#25443)

* Add area support to vacuum service schemas

* Fixed tests

* De-couple platform schemas

fredrike committed:
Daikin simplification and code cleanup (#25416)

* fix preset documentation

* Use pydaikin set holiday method

* update temperature readings, code simplification

* more temperature cleanup

* cleanup HVAC_MODE

* remove get() method and move code to respectivly place

* remove string constant in code

* remove get() method and move code to respectivly place

* isort results

* fixes in state method

zxdavb committed:
Bump geniushub client (#25458)

plafue committed:
Allow configuring sources for older Pioneer receivers (#25305)

* Allow configuring sources for older Pioneer receivers

* Replace config.get calls with dict[key] syntax

bachya committed:
Add area support to group service schemas (#25410)

scop committed:
Lint fixes (#25462)

scop committed:
Install requirements_test.txt for flake8 in Azure CI (#25463)

Was missing specified versions, as well as flake8-docstrings altogether.

Santobert committed:
Fix bloomsky unit system (#25460)

* initial commit - fix bloomsky unit system

* Add another warning

* Fix linting error

* Include metric sensor units

* Shorten a too long line

scop committed:
Type check homeassistant.scripts (#25464)

* Run mypy on homeassistant.scripts, disabling bunch of checks for now

* Declare async_initialize in AuthProvider

* Add some type hints

* Remove unreachable code

* Help mypy out

* Script docstring fixes

balloob committed:
Fix Nest turning off eco (#25472)

* Fix Nest turning off eco

* Add hvac action

scop committed:
Mypy config cleanups (#25475)

* Move file specific config to inline comments

* Disallow untyped defs by default everywhere

balloob committed:
Allow cors for static files (#25468)

michaeldavie committed:
Add scan interval to config of Environment Canada sensor (#25414)

* Make refresh period configurable

* Remove throttle, set SCAN_INTERVAL

* Move SCAN_INTERVAL to module

Kane610 committed:
UniFi block clients (#25478)

* Allow blocking clients on UniFi networks

Kane610 committed:
Add a unique identifier to deCONZ groups (#25485)

* Add a unique identifier to deconz groups

aschamberger committed:
Add ord() to template filters (#25398)

* Add ord() to template filters

* Remove trailing whitespace

* add test

MartinHjelmare commented on homeassistant/components/ring/switch.py:
```diff
@@ -0,0 +1,115 @@
+
+    def turn_on(self, **kwargs):
+        """Turn the siren on for 30 seconds."""
+        self._on = True
```
How fast is the feedback from calling `update`? The `update` method is called at the end of the service call after the actuator method call, for polling entities. If the feedback is fast enough, < 2 seconds, from the device we don't need to change state holding instance attribute here.

rossdargan replied:
I'm going to fix this as a second pull request. I'd like to get this released, and the fix for this will impact the rest of the ring.

MartinHjelmare commented on homeassistant/components/ring/switch.py:
```diff
@@ -0,0 +1,115 @@
+
+    def update(self):
+        """Get the siren status from ring"""
+        BaseRingSwitch.update(self)
```
```py
super().update()
```

MartinHjelmare commented on homeassistant/components/ring/switch.py:
```diff
@@ -0,0 +1,115 @@
+
+    def update(self):
+        """Get the light status from ring"""
+        BaseRingSwitch.update(self)
```
See above.

MartinHjelmare commented on homeassistant/components/ring/switch.py:
```diff
@@ -0,0 +1,115 @@
+
+
+class BaseRingSwitch(SwitchDevice):
+    """Represents a switch for controlling an aspect of a ring device"""
```
Represent a switch...

Please end with period.

rossdargan replied:
I used flake8 - rest should be fine I hope!

MartinHjelmare commented on homeassistant/components/ring/switch.py:
```diff
@@ -0,0 +1,115 @@
+class BaseRingSwitch(SwitchDevice):
+    """Represents a switch for controlling an aspect of a ring device"""
+    def __init__(self, hass, device, device_type):
+        self._device = device
```
Missing docstring.

MartinHjelmare commented on homeassistant/components/ring/switch.py:
```diff
@@ -0,0 +1,115 @@
+        return self._unique_id
+
+    def update(self):
+        """Get the siren status from ring"""
```
Missing period at the end.

MartinHjelmare commented on homeassistant/components/ring/switch.py:
```diff
@@ -0,0 +1,115 @@
+
+
+class SirenSwitch(BaseRingSwitch):
+    """Used to turn the ring cameras siren on and off"""
```
Missing period at the end.

MartinHjelmare commented on homeassistant/components/ring/switch.py:
```diff
@@ -0,0 +1,115 @@
+        self._device.siren = 1
+
+    def turn_off(self, **kwargs):
+        """Turn the siren off"""
```
Missing period at the end.

MartinHjelmare commented on homeassistant/components/ring/switch.py:
```diff
@@ -0,0 +1,115 @@
+        self._device.siren = 0
+
+    def update(self):
+        """Get the siren status from ring"""
```
Missing period at the end.

MartinHjelmare reviewed (commented):
Please rebase on latest dev branch to try and have flake8 check docstrings.

pvizeli committed:
Update azure-pipelines-ci.yml for Azure Pipelines

pvizeli committed:
Update azure-pipelines-ci.yml for Azure Pipelines

squishykid committed:
Update solax to 0.1.2 (#25497)

bollewolle committed:
Add De Lijn (Flemish Public Transport) component (#24265)

* Initial commit of De Lijn (Flemish Public Transport) component

* Code corrections as per HA dev's requests

* changes to variable naming, setting attribution and states, plus some smaller optimizations

* Overlooked some linting issues, these are now fixed

* Updated pydelijn version requirement to 0.5.1 so UTC timestamps can be used instead of relative/local times.
Removed unused definition

* Updated pydelijn version requirement to 0.5.1 in requirements_all.txt

* Update the self._attributes dict directly instead of replacing it
Assign ATTRIBUTION while creating the _attributes dict
Remove the ATTRIBUTION assignment in device_state_attributes as it's updated in the async_update now.

* Linting issue (lenght of 2 lines) solved

* Removed a relative time attribute
Updated a linting issue in the LOGGER (used % instead of the format)

pvizeli committed:
[skip ci] Update azure-pipelines-ci.yml for Azure Pipelines

pvizeli committed:
Update azure-pipelines-ci.yml for Azure Pipelines

pvizeli committed:
Update azure-pipelines-ci.yml for Azure Pipelines

pvizeli committed:
Update azure-pipelines-ci.yml for Azure Pipelines

pvizeli committed:
Update azure-pipelines-ci.yml for Azure Pipelines

pvizeli committed:
Update azure-pipelines-ci.yml for Azure Pipelines

pvizeli committed:
Update azure-pipelines-ci.yml for Azure Pipelines

rossdargan commented on homeassistant/components/ring/switch.py:
```diff
@@ -0,0 +1,115 @@
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up a sensor for a Ring device."""
+    ring = hass.data[DATA_RING]
+    logging.warning('setting up')
```
:/ oops.

rossdargan committed:
Merge branch 'dev' of https://github.com/rossdargan/home-assistant into dev

rossdargan commented:
ah man I've screwed this pull request with a bad merge. I will delete it and re-create. Sorry for everyone who must have been pinged by this!
